common/regex-partial.cpp

```diff
@@ -0,0 +1,203 @@
+#include "regex-partial.h"
+#include "common.h"
+#include <functional>
+#include <optional>
+
+common_regex::common_regex(const std::string & pattern) :
+    pattern(pattern),
+    rx(pattern),
+    rx_reversed_partial(regex_to_reversed_partial_regex(pattern)) {}
+
+common_regex_match common_regex::search(const std::string & input, size_t pos, bool as_match) const {
+    std::smatch match;
+    if (pos > input.size()) {
+        throw std::runtime_error("Position out of bounds");
+    }
+    auto start = input.begin() + pos;
+    auto found = as_match
+        ? std::regex_match(start, input.end(), match, rx)
+        : std::regex_search(start, input.end(), match, rx);
+    if (found) {
+        common_regex_match res;
+        res.type = COMMON_REGEX_MATCH_TYPE_FULL;
+        for (size_t i = 0; i < match.size(); ++i) {
+            auto begin = pos + match.position(i);
+            res.groups.emplace_back(begin, begin + match.length(i));
+        }
+        return res;
+    }
+    std::match_results<std::string::const_reverse_iterator> srmatch;
+    if (std::regex_match(input.rbegin(), input.rend() - pos, srmatch, rx_reversed_partial)) {
+        auto group = srmatch[1].str();
+        if (group.length() != 0) {
+            auto it = srmatch[1].second.base();
+            // auto position = static_cast<size_t>(std::distance(input.begin(), it));
+            if ((!as_match) || it == input.begin()) {
+                common_regex_match res;
+                res.type = COMMON_REGEX_MATCH_TYPE_PARTIAL;
+                auto begin = std::distance(input.begin(), it);
+                GGML_ASSERT(begin >= 0);
+                auto end = input.size();//begin + group.length();
+                GGML_ASSERT(static_cast<size_t>(begin) <= end);
+                res.groups.push_back({static_cast<size_t>(begin), end});
+                return res;
```

Comment:
This is maybe slightly cleaner:

```suggestion
                const size_t begin = std::distance(input.begin(), it);
                GGML_ASSERT(begin >= 0);
                const size_t end = input.size();//begin + group.length();
                GGML_ASSERT(begin <= end);
                res.groups.push_back(begin, end});
                return res;
```